### ISSUE TITLE
AutoConfiguredMoqCustomization not setting properties of abstract classes declared on interfaces

### ISSUE THREAD

**Issue by dcastro** [AutoFixture/AutoFixture#324]
This issue relates to the problem described here: [AutoConfiguredMoqCustomization with abstract class implementing interface][0].

**Background**

This customization uses the `MockSealedPropertiesCommand` command to set the value of a mocked object's non-overridable properties (overridable properties, on the other hand, are handled by `MockVirtualMethodsCommand`).

This class, in turn, uses `AutoFixture.Kernel`'s `AutoPropertiesCommand`.

Abstraction layers aside, the following code illustrates a simplification of what these classes do behind the scenes:

```csharp
Mock mock = new Mock<AbstractSutWithInterface>();

var specimen = mock.Object;

//retrieve sealed properties
var specimenType = specimen.GetType();
var sealedProperties = specimenType.GetProperties()
           .Where(p =>  ! p.GetSetMethod().IsVirtual || p.GetSetMethod().IsFinal)

//then set the properties
// ...
```

Notice that we're not actually retrieving the SUT's (e.g., `AbstractSutWithInterface`'s) properties! `specimenType` is castle's generated proxy - i.e., `Castle.Proxies.AbstractSutWithInterfaceProxy` -, **not** `AbstractSutWithInterface`.

**The problem**

There's an underlying assumption in the code above: `AbstractSutWithInterfaceProxy`'s properties should be the same as `AbstractSutWithInterface`'s properties. And if the SUT's properties are sealed, so should the proxy's properties be. In my opinion, this is a fair assumption.

However, since Moq v4.2.1408.619, this seems to have changed. 

Castle's proxy now not only inherits the SUT's properties, but also re-declares them *and* does not mark them as `final` as it should.

The test case:

```csharp

[Fact]
public void TestAbstractSutWithInterface()
{
    //print the SUT's properties
    PrintProperties(typeof(AbstractSutWithInterface));
    
    //print the proxy's properties
    var mock = new Mock<AbstractSutWithInterface>();
    PrintProperties(mock.Object.GetType());
}

public void PrintProperties(Type type)
{
    foreach (var set in type.GetProperties().Select(x => x.GetSetMethod()).Where(x => x != null))
    {
        Console.WriteLine("-------");
        Console.WriteLine("Setter Name: " + set.Name);
        Console.WriteLine("Declaring Type: " + set.DeclaringType);
        Console.WriteLine("IsVirtual: " + set.IsVirtual);
        Console.WriteLine("IsFinal: " + set.IsFinal);
    }
}
```

Output using Moq 4.2.1402.2112 (Last working version - 21 Feb 2014).
Here, we see the proxy's properties match the SUT's properties exactly, as expected:
```
-------
Setter Name: set_MigrationParameters
Declaring Type: Ploeh.AutoFixture.AutoMoq.UnitTest.AbstractSutWithInterface
IsVirtual: True
IsFinal: True
-------
Setter Name: set_MigrationParameters
Declaring Type: Ploeh.AutoFixture.AutoMoq.UnitTest.AbstractSutWithInterface
IsVirtual: True
IsFinal: True
```

Output using Moq 4.2.1408.619 (First non-working version - 6 Aug 2014).
```
-------
Setter Name: set_MigrationParameters
Declaring Type: Ploeh.AutoFixture.AutoMoq.UnitTest.AbstractSutWithInterface
IsVirtual: True
IsFinal: True
-------
Setter Name: set_MigrationParameters
Declaring Type: Castle.Proxies.AbstractSutWithInterfaceProxy <--- the property has been re-declared?
IsVirtual: True
IsFinal: False      <-- bug?
```

Here, we see that the proxy re-declared the property, and this declaration is not marked as `final` as it should - meaning this property is now effectively virtual, aka overridable. And since it is overridable, `MockSealedPropertiesCommand` ignores it.

This smells like a bug to me. Even if it was by design, it's a breaking change and should not have been introduced without upreving the major version number.

**My question to the owners and contributors is**: since we're not sure whether this is a bug or by design, should we work around this issue? What's the correct course of action here? How do we make sure whether this is a bug? Should I also report this as a bug on Moq's github project?

I've thought of a workaround, and it seems to be quick and easy. The idea is to call `GetProperties` on `AbstractSutWithInterface` directly, instead of on castle's proxy. To do this, `MockSealedPropertiesCommand` has to be changed from this:

```csharp
public class MockSealedPropertiesCommand : ISpecimenCommand
{
    private readonly ISpecimenCommand autoPropertiesCommand =
        new AutoPropertiesCommand(new MockSealedPropertySpecification());


    public void Execute(object specimen, ISpecimenContext context)
    {
        //...
        autoPropertiesCommand.Execute(mock.Object, context);
    }
}
```

To this:

```csharp
public class MockSealedPropertiesCommand : ISpecimenCommand
{
    public void Execute(object specimen, ISpecimenContext context)
    {
        //...

        var autoPropertiesCommand =
            new AutoPropertiesCommand(mock.GetType().GetMockedType(), new MockSealedPropertySpecification());
        autoPropertiesCommand.Execute(mock.Object, context);
    }
}
```

I've tested this and it solves the issue without breaking any regression tests.

  [0]: http://stackoverflow.com/q/26523960/857807

**Comment by dcastro** [comment]
This bug has been "fixed" as a consequence of PR #336.

To recap: a bug in Moq/Castle Proxy caused some sealed properties to be incorrectly marked as virtual by dynamic proxy. This would in turn cause `MockSealedPropertiesCommand` ignore these properties and not set them.

In #336, `MockSealedPropertiesCommand` was deprecated, and we now use `AutoMockPropertiesCommand` to set every property of the mock, regardless of whether they're overridable or not.

The bug in either Moq or Castle Proxy still exists, but it doesn't affect the functionality of `AutoConfiguredMoqCustomization` anymore.

**Comment by dcastro** [comment]
Here are the test cases, adapter from the post on StackOverflow:

```csharp
public interface IMigrationParameters
{
}

public interface IMigration
{
    IMigrationParameters MigrationParameters { get; set; }
}

public abstract class AbstractSutWithoutInterface
{
    public IMigrationParameters MigrationParameters { get; set; }
}

public abstract class AbstractSutWithInterface : IMigration
{
    public IMigrationParameters MigrationParameters { get; set; }
}

public class UnitTests
{
    [Fact]
    public void TestAbstractSutWithoutInterface()
    {
        var fixture = new Fixture();
        fixture.Customize(new AutoConfiguredMoqCustomization());

        var mock = fixture.Create<AbstractSutWithoutInterface>();

        Assert.NotNull(mock.MigrationParameters);
    }

    [Fact]
    public void TestAbstractSutWithInterface()
    {
        var fixture = new Fixture();
        fixture.Customize(new AutoConfiguredMoqCustomization());

        var mock = fixture.Create<AbstractSutWithInterface>();

        Assert.NotNull(mock.MigrationParameters);
    }
}
```
